### PR TITLE
Filter Panel: Location match amount display

### PR DESCRIPTION
**Pull request by AlexanderGeere**
## Description

A count of the locations that match the filter criteria will now reflect on the filter panel.
Screenshot 2024-11-04 112035

This can be hidden by supplying: `layer.filter.location_count.hidden = true`.

## GitHub Issue
[#1621](https://github.com/GEOLYTIX/xyz/issues/1621)

## Type of Change
- ✅ New feature (non-breaking change which adds functionality)

## How have you tested this? 
Tested locally on `bugs_testing/ui_elements/workspace.json`. Can tested on any layer with a filter. 

## Testing Checklist 
- ✅ Existing Tests still pass
- ✅ Ran locally on my machine

## Code Quality Checklist
- ✅ My code follows the guidelines of XYZ
- ✅ My code has been commented
- ✅ Documentation has been updated
- ✅ New and existing unit tests pass locally with my changes
- ✅ Main has been merged into this PR

### CODE REVIEW

File: lib/ui/layers/panels/filter.mjs [hunk 1]
```diff
@@ -15,6 +15,7 @@ mapp.utils.merge(mapp.dictionaries, {
     layer_filter_greater_than: 'Greater than',
     layer_filter_less_than: 'Less than',
     layer_filter_set_filter: 'Set filter',
+    layer_filter_location_match: 'Location(s) match your criteria'
```
Comment (RobAndrewHurst) [inline lib/ui/layers/panels/filter.mjs:18]: this is going to want to wait for the Dictionaries issue #1434 as all these merges are not happening in the core anymore

**Review by RobAndrewHurst — CHANGES_REQUESTED** [review]

File: lib/ui/layers/panels/filter.mjs (hunk 1 above)
Comment (AlexanderGeere) [reply lib/ui/layers/panels/filter.mjs]: I removed this for line for now.

File: lib/utils/locationCount.mjs [hunk 2]
```diff
@@ -0,0 +1,19 @@
+export async function locationCount(layer){
```
[the remark this answers is not in the document]
Comment (AlexanderGeere) [reply lib/utils/locationCount.mjs:21]: Done!

**Review by dbauszus-glx — DISMISSED** [review]
Modules should not have a default export and export.

This looks odd.
```js
import filter, { updateCount } from './panels/filter.mjs'
```

I'd just export the filterPanel and updateCount as regular export from the module.

```js
import { filterPanel, updateCount } from './panels/filter.mjs'
```

And then assign the filterPanel method to the filter method property in the panels{}

I wouldn't bother with layer.filter.location_count.hidden flag and just show this. We don't have flags to hide the clear or reset button.

The locationCount module should be in the ui utils not the general mapp.utils since this is only used by ui methods.

**Comment by RobAndrewHurst** [comment]
Is this something we want to be optional or even configurable?

**Comment by simon-leech** [comment]
> Is this something we want to be optional or even configurable?

I don't think so - I think it should be core like the Clear and Reset buttons. Its a helpful feature to provide for UX

**Comment by RobAndrewHurst** [comment]
> > Is this something we want to be optional or even configurable?
> 
> I don't think so - I think it should be core like the Clear and Reset buttons. Its a helpful feature to provide for UX

My thought behind this was to include custom count messages, and also the option to turn off if we have crazy data sets that we just don't want to include

**Review by RobAndrewHurst — APPROVED** [review]

**Review by dbauszus-glx — CHANGES_REQUESTED** [review]
I still want to look at this.

**Review by dbauszus-glx — APPROVED** [review]
This wants another review.

The updateCount method does not have to await the layer and does not need to return the layer. The method should not exported as layers.panel.updateCount which is meaningless.

The method is a filter method and should be assigned as mapp.ui.layers.filters.updateCount(layer)

This method should be called on the changeEnd event which is triggered by the applyFilter method.

The changeEnd event listener is assigned in the filterPanel method when we know that the view as well as the count element do exist.

The locationCount utility method should respect if the viewport flag is set in the layer.filter configuration. This enables viewport queries on layers such as MVT which do not have a viewport queryparams.

**Review by simon-leech — CHANGES_REQUESTED** [review]
Found a few little issues: 

1. When the layer is off and you expand the filter panel - you end up just seeing the 'Locations match your criteria' text. This should be hidden until you turn on the layer. 
image
2. When you turn on the layer, the number is not updated until you zoom in / out - this should be updated when you toggle display of the layer. 
3. I'm seeing a lot of errors in the console from the location_count query, it looks like the table parameter is not being passed to it. 
4. Can we update the documentation to make it clear that you need to pass 'viewport' to get the number updating when you zoom in / out - as this is basically required otherwise its pretty confusing for the user.

**Review by RobAndrewHurst — APPROVED** [review]

**Review by simon-leech — APPROVED** [review]
All good with me :)

**Pull request merged**